hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -110,8 +124,90 @@ public SqlResult execute(@Nonnull SqlStatement statement) {
             return res;
         } catch (Exception e) {
             RuntimeException error = rethrow(e, connection);
-            res.onExecuteError(error);
-            throw error;
+            SqlResubmissionResult resubmissionResult = resubmitIfPossible(res, error);
+            if (resubmissionResult == null) {
+                res.onExecuteError(error);
+                throw error;
+            }
+            handleExecuteResubmittedResponse(res, resubmissionResult);
+            return res;
+        }
+    }
+
+    @SuppressWarnings("BusyWait")
+    SqlResubmissionResult resubmitIfPossible(SqlClientResult result, RuntimeException error) {
+        if (!shouldResubmit(error) || !shouldResubmit(result)) {
+            return null;
+        }
+
+        long resubmissionStartTime = System.nanoTime();
+        int invokeCount = 0;
+
+        SqlResubmissionResult resubmissionResult = null;
+        do {
+            logger.info("Resubmitting query: " + result.getQueryId());
```

Comment:
Why do you want to avoid constructing the strings? Javac compiles sting concatenation to code equivalent to ```StringBuilder``` and to even better code since JDK9. The ```LoggingUtil``` uses ```String.fomat``` which has worse performance.

```java
public class StringConcat {
    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    public void builder(Blackhole blackhole) {
        StringBuilder builder = new StringBuilder();
        builder
                .append("a")
                .append(this.hashCode())
                .append("a");
        blackhole.consume(builder.toString());
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    public void concat(Blackhole blackhole) {
        String result = "a" + this.hashCode() + "a";
        blackhole.consume(result);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    public void stringFormat(Blackhole blackhole) {
        String result = String.format("a%da", this.hashCode());
        blackhole.consume(result);
    }
}
```

Results:
```
StringConcat.builder       avgt   25   36.943 ± 0.567  ns/op
StringConcat.concat        avgt   25   36.877 ± 0.362  ns/op
StringConcat.stringFormat  avgt   25  203.401 ± 9.362  ns/op
```

Reply:
You don't need to concat at all if the logging level is disabled, which is the case which we optimize for. We don't expect any customer to have FINE logging on, it will generate a huge amount of logs.

Another option would be to provide a supplier for the utility, but that would be GC litter. The current way it's the fastest if the logging is disabled. That's also the reason why we don't use varargs, as those also create more litter.